File: src/api/internal/models/datatemplate.go
// Package models describes the structure of the database models
package models

import "go.mongodb.org/mongo-driver/bson/primitive"

// DataTemplateColumn defines the structure of a column
type DataTemplateColumn struct {
	OriginalName string `bson:"originalName" json:"originalName"`
	PrettyName   string `bson:"prettyName" json:"prettyName"`
	DataType     string `bson:"dataType" json:"dataType"`
}

// DataTemplate defines the structure of a data table
type DataTemplate struct {
	// TODO: Update _id to id for json
	ID      primitive.ObjectID   `bson:"_id,omitempty" json:"_id,omitempty"`
	Name    string               `bson:"name" json:"name"`
	Columns []DataTemplateColumn `bson:"columns" json:"columns"`
	Files   []FileMetadata       `bson:"files,omitempty"`
}

File: src/Excalibur.Api/Models/DataTemplate.cs
﻿namespace Excalibur.Api.Models;

using Excalibur.Api.DTOs;
using Excalibur.Api.DTOs.Responses;
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

public class DataTemplate
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string? Id { get; set; }

    [BsonElement("name")]
    public string Name { get; set; } = string.Empty;

    [BsonElement("columns")]
    public List<DataTemplateColumn>? Columns { get; set; }

    [BsonElement("files")]
    public List<DataTemplateUploadedFileMetadata> Files { get; set; }

    public DataTemplateResponse MapToResponse()
    {
        return new DataTemplateResponse
        {
            Id = Id,
            Name = Name,
            Columns = Columns is null ? new List<DataTemplateColumnDto>() : Columns.Select(c => new DataTemplateColumnDto
            {
                Id = c.Id.ToString(),
                OriginalName = c.OriginalName,
                PrettyName = c.PrettyName,
                DataType = c.DataType
            }).ToList(),
        };
    }
}
